etcd-client/async_etcd_client.py

```diff
@@ -0,0 +1,219 @@
+        """
+        Parameters
+        ----------
+        host : str
+            The address of etcd
+        port : int
+            The port of etcd accepting client connections
+        """
+        self.host = host
+        self.port = port
+
+    async def get(self, key, prefix=False):
+        """Returns the value stored in etcd associated with a key or a
+        list of key-value pairs stored in etcd associated with keys
+        starting with a given prefix.
+
+        If the argument `prefix` isn't passed in, it won't treat the
+        key as a prefix.
+
+        Parameters
+        ----------
+        key : str
+            The key to search for or the prefix if prefix is True
+        prefix : bool, optional
+            Whether or not treat the given key as prefix and search for
+            all the keys starting with that prefix (default is None)
+
+        Raises
+        ------
+        etcd.EtcdKeyNotFound
+            If the key isn't found in etcd
+        """
+        client = etcd.Client(self.host, self.port)
+        try:
+            if prefix:
+                rv = []
+                resp = await client.read(key, recursive=True, sorted=True)
+                for child in resp.children:
+                    rv.append("%s: %s" % (child.key, child.value))
+                return rv
+            resp = await client.read(key)
+            return resp.value
+        except etcd.EtcdKeyNotFound as e:
+            client.close()
+            raise e
+
+    async def set(self, key, value, swap=False, oldvalue=""):
+        """Sets a new key-value pair in etcd and returns the value or
+        replaces a value associated with a key with a new one and
+        returns the new value.
+
+        If the argument `swap` isn't passed in, it won't try to replace
+        the value of the given key.
+
+        Parameters
+        ----------
+        key : str
+            The key to be added in etcd
+        value : str
+            The value of the new key
+        swap : bool, optional
+            Whether or not to swap `value` and `oldvalue` associated
+            with the given key (default is None)
+        oldvalue : str, optional
+            The value associated with the given key meant to be
+            replaced (default is "")
+
+        Raises
+        ------
+        etcd.EtcdAlreadyExist
+            If the key already exists in etcd and `swap` is False
+        etcd.EtcdCompareFailed
+            If `oldvalue` is not the value stored in etcd for the given
+            key
+        """
+        client = etcd.Client(self.host, self.port)
+        try:
+            if swap:
+                await client.write(key, value, prevValue=oldvalue)
+                return value
+            await client.write(key, value, prevExist=False)
+            return value
+        except etcd.EtcdAlreadyExist as e:
+            client.close()
+            raise e
+        except etcd.EtcdCompareFailed as e:
```

Comment:
you should be able to do 
```
except (etcd.EtcdAlreadyExist,etcd.EtcdCompareFailed)  as e:
   ...
```

To avoid having to repeat yourself.

Another question, i see a lot of 

```
try:
   ...
except ...:
   client.close()
```

Do you know if client can be use as a context manager? 

It also look to me that the client is not closed when the functions are exiting normally , is that on purpose ? 

I'm not familiar with etcd/aio_etcd so asking.

Reply:
From my understanding, unfortunately the client can't be used as a context manager.
The `aio_etcd` documentation doesn't even mention closing the client. However, I noticed that every time an exception was raised, I got an error saying that the client wasn't closed properly but I think you are right and I should close the client explicitly when the functions exit normally also.